common/tools/dev-tool/src/commands/run/update-snippets.ts

```diff
@@ -0,0 +1,530 @@
+  {},
+);
+
+const log = createPrinter("update-snippets");
+
+const SNIPPET_PATH = ["test", "snippets.spec.ts"];
+
+/**
+ * Describes a location where a snippet is actually presented to a reader, i.e. a Markdown file or JSDoc comment
+ * location.
+ */
+interface SnippetLocationInfo {
+  /** The snippet name determined by the tag on its enclosing fence. */
+  name: string;
+  /** The fence string that started this snippet. */
+  fence: string;
+  /** The target language of the snippet. */
+  language: "js" | "ts";
+  /** The absolute path to the file the snippet is located in. */
+  absoluteFilePath: string;
+  /** The contents of the snippet with the line prefix stripped */
+  lineTrimmedContents: string[];
+  /** The range of lines in the file where the snippet appears. */
+  lineRange: [number, number];
+  /** The prefix that appeared before the code fence, which will be inserted before each line in the snippet. */
+  linePrefix: string;
+}
+
+/**
+ * Finds all candidate files that _might_ contain snippet locations.
+ *
+ * @param dir - the base directory of the package
+ */
+async function* getAllSnippetFiles(dir: string): AsyncIterable<string> {
+  // Only consider markdown files up to a depth of 1 (i.e. _in_ the project folder). This is to prevent grabbing things
+  // like samples/*/README.md and other similar files that are not really part of the "source" of the package.
+  yield* findMatchingFiles(dir, (name) => name.endsWith(".md"), { maxDepth: 1 });
+
+  yield* findMatchingFiles(path.join(dir, "src"), (name) => name.endsWith(".ts"));
+
+  yield* findMatchingFiles(path.join(dir, "test"), (name) => name.endsWith(".ts"));
+
+  if (existsSync(path.join(dir, "samples-dev"))) {
+    yield* findMatchingFiles(path.join(dir, "samples-dev"), (name) => name.endsWith(".ts"));
+  }
+}
+
+/**
+ * Finds all the snippet locations in a project.
+ *
+ * @param info
+ * @returns
+ */
+async function findAllSnippetLocations(info: ProjectInfo): Promise<SnippetLocationInfo[]> {
+  const snippets: SnippetLocationInfo[] = [];
+  let hadError = false;
+
+  for await (const f of getAllSnippetFiles(info.path)) {
+    // We want to use some kind of semantically appropriate comment to allow ignoring a file.
+    const ignoreString = f.endsWith(".md")
+      ? "<!-- dev-tool snippets ignore -->"
+      : "// dev-tool snippets ignore";
+
+    const contents = (await fs.readFile(f)).toString("utf-8");
+
+    if (contents.includes(ignoreString)) {
+      log.info(`Ignoring file ${path.relative(info.path, f)} per file comment directive.`);
+      continue;
+    }
+
+    let openFence: [RegExpMatchArray, number, string[]] | undefined = undefined;
+
+    for (const [line, idx] of contents.split(/\r?\n/).map((line, idx) => [line, idx] as const)) {
+      // We'll scan the file for matching code fences manually.
+      if (openFence === undefined) {
+        // We allow fences to be nested up to an arbitrary depth, but practically speaking we will never have fences
```

Comment:
Maybe we should enforce that from the beginning which will hopefully simplify the regex? Just a thought, not a big deal

Reply:
Probably, but this is good for a first pass.  If this becomes an issue, we can revisit.